Add QueryFactory class.

ADmad opened the pull request:
Refs #16648

ADmad committed:
Add QueryFactory class.

markstory commented on src/Database/Connection.php:
```diff
@@ -278,44 +280,74 @@ public function run(Query $query): StatementInterface
+        ExpressionInterface|Closure|array|string|float|int $fields = [],
+        array|string $tables = [],
+        array $types = []
+    ): SelectQuery {
```
I was initially concerned about all these positional parameters, but users can also used named parameters :smile: Perhaps we should document the named parameter API instead of the positional one.

ADmad replied:
> Perhaps we should document the named parameter API instead of the positional one.

How do you propose we do that? The tools will start complaining if the params aren't documented the "standard" way.

markstory replied:
> How do you propose we do that?

I was thinking of the examples we provide in the book and in the 'example' blocks in the API docs. I don't think we should change much in how we define types/params in doc blocks.

ADmad replied:
I guess we could do that.

othercorey commented on src/Database/Connection.php:
```diff
@@ -278,44 +280,74 @@ public function run(Query $query): StatementInterface
-        return new SelectQuery($this);
+    public function newSelectQuery(
+        ExpressionInterface|Closure|array|string|float|int $fields = [],
+        array|string $tables = [],
```
Should we call this `from` instead of `tables`? It's how you would read/write the SQL

markstory replied:
That's a great idea. It makes the named parameter interface more like the underlying methods which is a nice memory aid.

markstory replied:
Hrm. After doing all these changes and updating the tests I came to realize that the consistency of always having `table` or `table` was easier to operate than having `into` and `from` and more. In fact, I think we could just normalize on table and be better off.

othercorey replied:
> Hrm. After doing all these changes and updating the tests I came to realize that the consistency of always having `table` or `table` was easier to operate than having `into` and `from` and more. In fact, I think we could just normalize on table and be better off.

The one place that's not true is joining on select queries or selecting from CTE.

ADmad replied:
I agree that `table` doesn't feel right for SELECT queries since the `FROM` expression doesn't necessarily have to be a table.

markstory replied:
A subquery or CTE expression result is effectively a 'table' :wink:

othercorey replied:
I'm ok with using table in our nomenclature. It allows for some more flexibility in other areas.

ADmad replied:
I too can live with table :)

othercorey commented on src/Database/Connection.php:
```diff
@@ -278,44 +280,74 @@ public function run(Query $query): StatementInterface
      * Create a new InsertQuery instance for this connection.
      *
+     * @param string|null $table The table to insert rows into.
+     * @param array $values Associative array of column => value to be inserted.
```
Does this just represent a single row or can it be an array of VALUES()? It sounds like only supports a single row with a subset of column names.

ADmad replied:
Yes it only supports a single row. I copied the API from the current `Connection::insert()` method.

ADmad replied:
We can change the method to have separate $columns and $values arguments.

markstory replied:
Do we need a high-level API on QueryFactory to bulk insert records? The query builder already has methods for doing those operations.

ADmad replied:
> Do we need a high-level API on QueryFactory to bulk insert records?

Nope, I prefer keeping the `QueryFactory` simple.

markstory committed:
Use $table consistently.

Having one parameter name is simpler and easier to remember.

markstory committed:
Add tests that cover named parameter public API

If we are going to document using named parameters we need to have tests
for it or we will eventually regress on this.

markstory reviewed (commented):
I pushed a few commits up to make the parameter names consistent and updated some existing tests to use the named parameter API we are thinking of documenting.

ADmad commented:
Anything more to be done here?
